ksy_schema.json

```diff
@@ -241,6 +241,10 @@
         "encoding": {
           "type": "string"
         },
+        "pad-right": {
+          "type": "integer",
+          "description": "specify a byte which is the string or byte array padded with up to the total size"
```

Comment:
I think this should also explain:

* what the relationship with `terminator` is (i. e. what does it mean when you have `pad-right` and `terminator` set to different values?)
* whether this information is used for validation while parsing, or if it will only be used in the future for serialization

I don't know the answer to either question off the top of my head - you would probably need to check the kaitai-struct-compiler source code.

Reply:
Good question! The behavior of combining terminator and pad-right is demonstrated by [test str_pad_term](https://github.com/kaitai-io/kaitai_struct_tests/blob/574a31d4ff6e51f0c447140a75af38e25af3486f/formats/str_pad_term.ksy#L14), but there is only a field with fixed size:

```yaml
meta:
  id: str_pad_term
  encoding: UTF-8
seq:
  # ...
  - id: str_term_and_pad
    type: str
    size: 20
    terminator: 0x40
    pad-right: 0x2b
```

This compiles into something like:
```java
  this.strTermAndPad = KaitaiStream.bytesTerminate(KaitaiStream.bytesStripRight(this._io.readBytes(20), (byte) 43), (byte) 64, false);
```

So it
 1. reads 20 bytes (size)
 2. strips all occurences of byte 43 from the end until some other byte is reached (pad-right)
 3. strips everything after first occurence of byte 64 (terminator)

As you can see, the pad-right is actually pointless for parsing if proper terminator byte is present after the end of value, because terminator and pad-right are set to different values. It would make difference if terminator is missing (only padded - not well formed data) or if terminator = pad-right and include: true (it would be first right-stripped, so no terminator byte would be present anymore). It will be useful for serialization, of course.

This is quite simple example, just reading some fixed number of bytes and then just cutting some rubbish from the end.

I was wondering, what happens if no size is present (and the end is determined by terminator instead). This yields:

```javascript
  this.strTermAndPad = this._io.readBytesTerm(64, false, true, true);
```

So the pad-right is completely ignored. I think the expected behavior is to at first applying terminator and then the pad-right on the resulting value. I will take care of updating tests and fixing the compiler.

Reply:
> I think the expected behavior is at first applying terminator and then the pad-right on the resulting value.

Just to be clear - how would you expect a string with both `terminator` and `pad-right` to look in the byte stream? The compiler currently generates code for the order "contents, terminator, padding", and it sounds like you're expecting "contents, padding, terminator" instead.

I think the compiler's current interpretation makes more sense - the string contents are terminated by the `terminator` byte, and after that, bytes with the value from `pad-right` are used to pad the string up to the required length.

I agree though that the current code generation is suboptimal. When both `terminator` and `pad-right` are specified, `pad-right` should be ignored when reading, and only the `terminator` should be used to detect where the string ends. I also think that `pad-right` shouldn't be allowed on variable-width string fields at all - the padding feature (as ksc currently interprets it) only makes sense for fixed-width fields.

Reply:
> The compiler currently generates code for the order "contents, terminator, padding", and it sounds like you're expecting "contents, padding, terminator" instead.

Sorry, you're right, of course. I was confused by the fact, that it doesn't make any sense using `pad-right` on a variable-width string (or byte array), as you mention later, so I incorrectly fixed my opinion so it does :-)
I agree that KSC should fail in this case.

So what shall we write into the description of `pad-right` in the schema? We should definitely state here that it can be used only with `size`. (It would be also nice to force this in JSON schema.) Have you something else in mind?

Reply:
I think the most important points (in addition to what you already wrote) are:

* If `terminator` is not specified, then `pad-right` controls where the string ends (basically it becomes the terminator).
* If `terminator` is specified, the padding comes after the terminator, not before. This means that in this case `pad-right` is only relevant for serialization and has no effect on parsing.